### ISSUE TITLE
Support for larger __init__py module addons, where dynamic use of importlib should not be present

### ISSUE THREAD

**Issue by StefanKarlsson987** [robertguetzkow/blender-python-examples#6]
The current example shows only how to get the modules loaded dynamically. It doesnt provide example of how to import other files where this dynamic loading should not be present.

For more elaborate addons, the __init__.py could handle the dynamic installing and loading of the dependencies. However, this would mean that the __init__.py file can only import the other py-files dynamically, after the loading and installing is done. This poses a few easily overcome challenges that needs to be fixed. I have done this already in my personal project addon, and could extract this into an example form for this repo when time permits

**Comment by robertguetzkow** [comment]
I'm not sure if this qualifies as a bug, the add-on is only a simple single file example. If you want to create an add-on that is spread across multiple files, this will of course need adjustments. 

What changes would you suggest?

**Comment by StefanKarlsson987** [comment]
In my addon based on the example, I have no importlib imported in any file other then the __init__.py. I think that would be nice to do. A major reason for wanting to add libs is to easily include code you have made in other projects. Having to change all the imports to dynamic calls is just work you dont want to do.

This example could be extended into 2 files, an init file that is much like the current example, and an extra file.py, that contains the dummy operator... however the secondary file should just do 

import lib as local

and use local as ref

**Comment by robertguetzkow** [comment]
I don't see how the current add-on suggests that every import has to be done programmatically. The way you've described the solution is exactly how you should be implementing this. Only the dependencies that don't come with Blender are installed and loaded through the provided functions and only once in the `__init__.py`.

An example that demonstrates this would perhaps be helpful. Maybe I've misunderstood you but the current implementation doesn't do anything wrong in your opinion, correct?

**Comment by robertguetzkow** [comment]
I see that the comment in the source code could cause a misunderstanding. I will update it.

**Comment by StefanKarlsson987** [comment]
With the current addon, if you want to import a dependency like this:

import matplotlib

you cannot. The addon will crash immediately on first load, never reaching the commands that installs and dynamically imports.
After you have succesfully run the addon once, you could modify the code to put the static import command in, and the addon would run fine.

This is because the import command is evaluated statically.
My way around this was pretty straightforward. Let the __init__.py file only handle the dynamic imports. In that file, only dependencies that are part of the Blender bundled python are allowed to be imported statically.

Only if the addon succesfully succeeds in dynamically importing the dependencies, will the __init__.py file import (also dynamically!) all of the files that are in the same folder structure as the __init__.py file.

This will allow for the imported files to use the static import, because they will only be interpreted as the dynamic import is calling them.

A small modification to the typical convention in addon structure was the need to encapsulate the "classes" list in a function. This is because the interpreter would otherwise evaluate those class names as well now statically. The function "classes()" just returns the list, but the _init__.py file must make sure to not invoke the function until after the dynamic load of the files of the addon in the folder structure.

**Comment by StefanKarlsson987** [comment]
As I implemented this in my addon(from scratch, but using the technique I learnt from your example), which is work in progress for non-publically available software, I cannot share the code (it is mixed in with other stuff). As soon as I have alittle time over, I will seperate the relevant funtionality (the dynamic installing/importing across files) and share it with you, maybe as another pull request.

What I would really like to see happen is for this to continue to grow, and become a feature in Blender. An addon, or just a large available example, on how to install any and all dependencies you want, and only possibly breaking other addons (not blender itself). This would really mature the python interface in blender... and is a step towards making the scripting tab more like a proper environment... like jupyter etc. I think this could go a long way in that direction, even inside the current blender. The next big step would then be to get virtualenv properly running... all under the hood with similar techniques. 

maybe all this is too optimistic, but its fun

**Comment by robertguetzkow** [comment]
@StefanKarlsson987 Thank you for your input. I'll take another look at this, when I got some time. I'm currently quite busy working on the bug tracker and on my thesis, so this may take a while.

**Comment by robertguetzkow** [comment]
I've done a [quick test](https://github.com/robertguetzkow/blender-python-examples/tree/2020-12-02-multi-experimental/add-ons/install-dependencies-multi-file) that I've pushed to a separate branch. If you import the other parts of the add-on after the installation is done this seems to work fine and only requires minimal changes.

**Comment by StefanKarlsson987** [comment]
looked at it now, without having time to test it. 

It should work. Putting the register and unregister functions in the included files are different from my approach. It looks nice in the example init file, and means you can avoid my uglier solution of putting the list of classes to register in a function call.

However...

It means you have 2 functions, register and unregister... that if you call either one of them outside of the init.py file, it will cause a crash of the addon. Would make sense to combine our two approaches:

have one single function in example.py... called maybe... "Blender_Classes()", that just returns a list of the classes to be registered/unregistered. That way, the init file would still be the only place where the dangerous calls (that should be done just once) take place.

**Comment by robertguetzkow** [comment]
There are more things to be improved and checked. This was just a quick test I've implemented very late at night.
